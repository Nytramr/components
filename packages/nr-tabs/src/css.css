:host {
  display: block;
}

.tab-button {
  text-decoration: none;
  color: black;
  background-color: #cccccc;
  padding: 4px 8px;
  border-top: 1px solid black;
  border-left: 1px solid black;
  border-right: 1px solid black;
  border-top-left-radius: 5px;
  border-top-right-radius: 5px;
  display: block;
}

.tab-button.active {
  background-color: white;
}

#tabs-header {
  display: flex;
}

#tabs-body {
  position: relative;
  border: 1px solid black;
  flex-grow: 1;
}

::slotted(div) {
  position: absolute;
  display: none;
  width: 100%;
}

::slotted(div.active) {
  display: block;
}

.tab.active {
  display: block;
}

#tabs {
  display: flex;
  flex-direction: column;
  height: 100%;
}
