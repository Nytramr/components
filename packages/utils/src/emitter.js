/**
 * we need to use symbols until JS implements private properties
 */

const subscribers = Symbol('subscribers');

export class Emitter {
  constructor() {
    this[subscribers] = [];
    this['emit'] = this.emit.bind(this);
    this['addSubscriber'] = this.addSubscriber.bind(this);
    this['removeSubscriber'] = this.removeSubscriber.bind(this);
  }

  emit(payload) {
    this[subscribers].forEach((subscriber) => subscriber(payload));
  }

  addSubscriber(subscriber) {
    if (!this[subscribers].includes(subscriber)) {
      this[subscribers].push(subscriber);
    }
  }

  removeSubscriber(subscriber) {
    const index = this[subscribers].indexOf(subscriber);
    if (index >= 0) {
      this[subscribers].splice(index, 1);
    }
  }
}
