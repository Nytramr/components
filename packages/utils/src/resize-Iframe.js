export function resizeIframe(iframe) {
  let keepGoing = true;
  function resize() {
    if (keepGoing && iframe) {
      var dom = iframe.contentDocument || iframe.contentWindow.document;
      if (dom) {
        var height = dom.scrollingElement.scrollHeight || 800; // This default is because of the lightbox.
        iframe.height = height;
      }
      window.requestAnimationFrame(resize);
    }
  }

  window.requestAnimationFrame(resize);

  return function stop() {
    keepGoing = false;
  };
}

export default resizeIframe;
