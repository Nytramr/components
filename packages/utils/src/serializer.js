export class Serializer {
  static stringify(obj) {
    const instance = new Serializer();
    return instance.stringify(obj);
  }

  constructor() {
    this.references = new Map();
    this['replacer'] = this.replacer.bind(this);
  }

  serialize(obj) {
    const type = (typeof obj).toString();

    const name = obj.constructor ? obj.constructor.name : 'Null';

    return `${type} ${name}`;
  }

  replacer(_, value) {
    const type = (typeof value).toString();
    if (type === 'object') {
      if (this.references.has(value)) {
        return this.references.get(value);
      }
      const name = value.constructor.name;

      switch (name) {
      }
    }
    return value;
  }

  stringify(obj) {
    return JSON.stringify(obj, this.replacer);
  }
}
