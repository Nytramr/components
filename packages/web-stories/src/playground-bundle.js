import defineIframe from '@nytramr/nr-iframe';
import defineTabs from '@nytramr/nr-tabs';

defineTabs('web-tabs');
defineIframe('web-iframe-container');

// document.addEventListener('load-content', (event) => console.log('load', event, event.target));
document.addEventListener('DOMContentLoaded', (event) => {
  console.log('DOMContentLoaded', event);

  const iFrames = document.querySelectorAll('web-iframe-container');

  Array.from(iFrames).forEach((iFrame) => {
    iFrame.addEventListener('load-content', (event) => {
      console.log('load content', event, event.target);
      const Bridge = event.target.Bridge;
      if (Bridge) {
        const instance = new Bridge();
        instance.init();
      } else {
        console.log('no bridge');
      }
    });

    // if (Bridge) {
    //   const instance = new Bridge();
    //   instance.init();
    // } else {
    //   console.log('no bridge');
    // }
  });
});
