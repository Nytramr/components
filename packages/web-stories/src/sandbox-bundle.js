import IFrameBridge from './sandbox/iframe-bridge';

document.domain = 'localhost';
class MyBridge extends IFrameBridge {
  constructor(...args) {
    super(...args);
    this.eventHandlers = [];
  }

  notifyEvent(data) {
    this.sendEvent(IFrameBridge.events.PRINT_EVENT, data);
  }

  subscribeEvents(events) {
    // unsubscribe first
    this.eventHandlers.forEach(({ element, eventHandler }) => element.removeEventListener(eventHandler));

    // Events with selector - add event to that selector query result
    events
      .filter((event) => event.selector)
      .forEach(({ eventName, selector }) => {
        const eventHandler = (e) => {
          this.notifyEvent({
            eventName,
            event: e,
          });
        };
        const elements = document.querySelectorAll(selector);
        elements.forEach((element) => {
          element.addEventListener(eventName, eventHandler);
          this.eventHandlers.push({ element, eventHandler });
        });
      });

    // Events without selector - add events to the document
    events
      .filter((event) => !event.selector)
      .forEach(({ eventName }) => {
        const eventHandler = (e) => {
          this.notifyEvent({
            eventName,
            event: e,
          });
        };
        document.addEventListener(eventName, eventHandler);
        this.eventHandlers.push({ document, eventHandler });
      });
  }

  init({ events }) {
    console.log('Bridge init');
    this.subscribeEvents(events);
  }
}

window.PUBLIC = {
  ...window.PUBLIC,
  IFrameBridge: MyBridge,
};
