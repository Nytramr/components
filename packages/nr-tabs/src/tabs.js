import css from './css.css';

export class Tabs extends HTMLElement {
  constructor() {
    super();
    const shadowRoot = this.attachShadow({ mode: 'open' });
    shadowRoot.innerHTML = `
    <style>
      ${css}
    </style>
    <div id="tabs">
      <div id="tabs-header"></div>
      <div id="tabs-body">
        <slot></slot>
      </div>
    </div>`;

    this.tabHeader = shadowRoot.querySelector('#tabs-header');
    this._slot = shadowRoot.querySelector('slot');

    this.slotChangeHandler = this._slotChangeHandler.bind(this);
    this.changeTabHandler = this._changeTabHandler.bind(this);
  }

  _slotChangeHandler() {
    const tabElements = this._slot.assignedElements();
    const tabButtons = tabElements.map(this._createTabButton).join('');
    this.tabHeader.innerHTML = tabButtons;
  }

  _changeTabHandler(event) {
    event.stopPropagation();
    const tabId = event.target.dataset.tabid;

    if (tabId) {
      this.setActiveTab(tabId);
    }
  }

  _createTabButton(tabElement) {
    const tabId = tabElement.id;
    const tabLabel = tabElement.dataset.tablabel;
    const isActive = tabElement.classList.contains('active');

    return `<a href="#" part="tab-button ${isActive ? 'active' : ''}" class="tab-button ${
      isActive ? 'active' : ''
    }" data-tabId="${tabId}">${tabLabel || tabId}</a>`;
  }

  setActiveTab(tabId) {
    const tabButton = this.shadowRoot.querySelector(`[data-tabid=${tabId}]`);
    const tab = this.querySelector(`#${tabId}`);

    const activeButton = this.shadowRoot.querySelector('.tab-button.active');
    const activeTab = this.querySelector('.active');

    activeButton && activeButton.classList.remove('active');
    activeButton && activeButton.part.remove('active');
    activeTab && activeTab.classList.remove('active');

    tabButton && tabButton.classList.add('active');
    tabButton && tabButton.part.add('active');
    tab && tab.classList.add('active');
  }

  connectedCallback() {
    this._slot.addEventListener('slotchange', this.slotChangeHandler);
    this.tabHeader.addEventListener('click', this.changeTabHandler);
  }

  disconnectedCallback() {
    this._slot.removeEventListener('slotchange', this.slotChangeHandler);
    this.tabHeader.removeEventListener('click', this.changeTabHandler);
  }
}

export function define() {
  customElements.define('nr-tabs', Tabs);
}
