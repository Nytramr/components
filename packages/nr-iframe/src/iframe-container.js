/**
 * Custom element to create and manage synchronous communication with an Iframe, without using post messages
 *
 * The page loaded on the Iframe must have a global bridge class on its global scope.
 *
 * This custom element will throw custom events into the parent DOM on behalf of the Bridge instance inside the Iframe.
 *
 */

// import { retry } from '../../utils/ready';
import { resizeIframe } from '@nytramr/utils/lib/resize-Iframe';

document.domain = 'localhost';

class IFrameContainer extends HTMLElement {
  static get observedAttributes() {
    return ['src'];
  }

  constructor() {
    super();
    const root = this.attachShadow({ mode: 'open' });
    root.innerHTML = `
      <style>
        :host {
          display: block;
        }
        iframe {
          width: 100%;
          box-sizing: border-box;
          border: none;
        }
      </style>
    `;

    this._loadIframeHandler = this._loadIframeHandler.bind(this);
  }

  attributeChangedCallback(name, oldValue, newValue) {
    if (name !== 'src' && oldValue === newValue) {
      return;
    }

    switch (name) {
      case 'src':
        this._loadIframe(newValue);
        break;
    }
  }

  _loadIframeHandler() {
    console.log('handle');
    this.stopResizing = resizeIframe(this.iFrame);
  }

  _unloadIframe() {
    if (this.iFrame) {
      this.iFrame.removeEventListener(this._loadIframeHandler);
      this.stopResizing && this.stopResizing();
    }
  }

  _loadIframe(src) {
    /**
     * Stop current Iframe connection attempt (if any) or ignore it all together
     * remove current iFrame
     * show loading state? Maybe the children?
     * create a new iFrame with the given url
     * try to connect with the bridge
     *  If success -> Show Iframe
     *  On fail -> Show error (special slot?)
     */

    this._unloadIframe();
    const iFrame = document.createElement('iframe');
    iFrame.src = src;
    iFrame.addEventListener('load', this._loadIframeHandler);

    this.iFrame = iFrame;
    this.shadowRoot.appendChild(iFrame);
  }

  disconnectCallback() {
    this._unloadIframe();
  }
}

export default function define(tagName) {
  customElements.define(tagName, IFrameContainer);
}
