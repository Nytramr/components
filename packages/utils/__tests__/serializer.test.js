import { Serializer } from '../src/serializer';

describe('@nytramr/utils', () => {
  describe('Serializer', () => {
    describe('stringify', () => {
      describe('with no circular references', () => {
        test.each`
          title                   | object
          ${'Object'}             | ${{ a: 1 }}
          ${'Object with Array'}  | ${{ a: [1, 2, 3] }}
          ${'Object with Object'} | ${{ a: { b: 1 } }}
          ${'Array'}              | ${[1, 2, 3]}
          ${'Array with Array'}   | ${[1, 2, 3, ['a', 'b', 'c']]}
        `('returns $title', ({ object }) => {
          expect(Serializer.stringify(object)).toMatchSnapshot();
        });
      });
    });

    describe('serialize', () => {
      describe('with no circular references', () => {
        test.each`
          title                           | object                       | result
          ${'Object'}                     | ${{ a: 1 }}                  | ${'object Object'}
          ${'Object instance'}            | ${new Object()}              | ${'object Object'}
          ${'Object with no constructor'} | ${Object.create(null)}       | ${'object Null'}
          ${'Array'}                      | ${[1, 2, 3]}                 | ${'object Array'}
          ${'String'}                     | ${'hello world'}             | ${'string String'}
          ${'Set'}                        | ${new Set([1, 2, 1, 3])}     | ${'object Set'}
          ${'Map'}                        | ${new Map([[1, 2], [1, 3]])} | ${'object Map'}
          ${'Boolean'}                    | ${true}                      | ${'boolean Boolean'}
          ${'Number'}                     | ${100}                       | ${'number Number'}
          ${'Function'}                   | ${() => {}}                  | ${'function Function'}
          ${'Serializer class'}           | ${Serializer}                | ${'function Function'}
          ${'Serializer instance'}        | ${new Serializer()}          | ${'object Serializer'}
          ${'Serializer instance'}        | ${new Serializer()}          | ${'object Serializer'}
        `('returns $title', ({ object, result }) => {
          const serializer = new Serializer();
          expect(serializer.serialize(object)).toEqual(result);
        });
      });
    });
  });
});
