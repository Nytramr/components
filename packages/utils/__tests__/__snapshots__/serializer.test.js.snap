// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`@nytramr/utils Serializer stringify with no circular references returns Array 1`] = `"[1,2,3]"`;

exports[`@nytramr/utils Serializer stringify with no circular references returns Array with Array 1`] = `"[1,2,3,[\\"a\\",\\"b\\",\\"c\\"]]"`;

exports[`@nytramr/utils Serializer stringify with no circular references returns Object 1`] = `"{\\"a\\":1}"`;

exports[`@nytramr/utils Serializer stringify with no circular references returns Object with Array 1`] = `"{\\"a\\":[1,2,3]}"`;

exports[`@nytramr/utils Serializer stringify with no circular references returns Object with Object 1`] = `"{\\"a\\":{\\"b\\":1}}"`;
